cognite/client/_api/files.py

```diff
@@ -596,6 +597,143 @@ def upload_bytes(
+                >>> client.files.complete_multipart_upload(session)
+        """
+        file_metadata = FileMetadata(
+            name=name,
+            external_id=external_id,
+            source=source,
+            mime_type=mime_type,
+            metadata=metadata,
+            directory=directory,
+            asset_ids=asset_ids,
+            data_set_id=data_set_id,
+            labels=labels,
+            geo_location=geo_location,
+            source_created_time=source_created_time,
+            source_modified_time=source_modified_time,
+            security_categories=security_categories,
+        )
+        try:
+            res = self._post(
+                url_path=self._RESOURCE_PATH + "/initmultipartupload",
+                json=file_metadata.dump(camel_case=True),
+                params={"overwrite": overwrite, "parts": parts},
+            )
+        except CogniteAPIError as e:
+            if e.code == 403 and "insufficient access rights" in e.message:
+                dsid_notice = " Try to provide a data_set_id." if data_set_id is None else ""
+                msg = f"Could not create a file due to insufficient access rights.{dsid_notice}"
+                raise CogniteAuthorizationError(message=msg, code=e.code, x_request_id=e.x_request_id) from e
+            raise
+
+        returned_file_metadata = res.json()
+        upload_urls = returned_file_metadata["uploadUrls"]
+        upload_id = returned_file_metadata["uploadId"]
+
+        return FileMultipartUploadSession(FileMetadata._load(returned_file_metadata), upload_urls, upload_id)
+
+    def upload_multipart_part(
+        self,
+        upload_url: str,
+        content: str | bytes | TextIO | BinaryIO,
+        mime_type: str | None = None,
+    ) -> None:
+        """Upload part of a file to an upload URL returned from `begin_multipart_upload`.
+        Note that if `content` does not somehow expose its length, this method may not work
+        on Azure. See `requests.utils.super_len`.
+
+        Args:
+            upload_url (str): URL to upload file chunk to.
+            content (str | bytes | TextIO | BinaryIO): The content to upload.
+            mime_type (str | None): Optional mime type for the `Content-Type` header.
+        """
+        if isinstance(content, str):
+            content = content.encode("utf-8")
+
+        headers = {"Content-Type": mime_type}
+        upload_response = self._http_client_with_retry.request(
+            "PUT",
+            upload_url,
+            data=content,
+            timeout=self._config.file_transfer_timeout,
+            headers=headers,
+        )
+        if not upload_response.ok:
+            raise CogniteFileUploadError(
+                message=upload_response.text,
+                code=upload_response.status_code,
+            )
+
+    def complete_multipart_upload(self, session: FileMultipartUploadSession) -> None:
```

Comment:
After the complete_multipart_upload call is completed, the client normally would need to also check for if the file is marked as uploaded=true.

To simplify this checking for the client, the complete_multipart_upload could have a optional setting, for example: `wait_for_upload`, which will call the file `retrieve` operation on the fileID, and repeatedly check (with some backoff delay) until it sees the file is marked as uploaded = true. Normally this is done within 1 second after complete_multipart_upload has returned. Sometimes it can take several seconds though.

Reply:
Should this method return a boolean, to return status about that the uploaded flag is true or not?
Or just always true, as long as the response is a 200 or 201.

Reply:
Would they need to check that? Or would that be something they _could_ check if they wanted to download the file immediately. Generally when you upload files you don't download them immediately.

Reply:
I'll check this in the test, but for now I don't think it makes sense to check it in `complete`

Reply:
It is perhaps only for certain use cases it would be important to know that the file is fully persisted, and can be downloaded.

The suggestion I had can be added later if we see a real need.